packages/diagram/lib/diagram.dart

```diff
@@ -0,0 +1,442 @@
+    // Short circuit the begin frame to use our timestamp instead of the "real" one.
+    // Re-use the previous onBeginFrame so that all of the hot reload, etc. works.
+    dynamic oldBeginFrame = ui.window.onBeginFrame;
+    ui.window.onBeginFrame = (Duration timestamp) {
+      oldBeginFrame(_timestamp);
+    };
+  }
+
+  Duration _timestamp = Duration.zero;
+  GlobalKey _boundaryKey = new GlobalKey();
+
+  /// Determines the ratio between physical units and logical units.
+  ///
+  /// The [pixelRatio] describes the scale between the logical pixels and the
+  /// size of the output image. It is independent of the
+  /// [window.devicePixelRatio] for the device, so specifying 1.0 (the default)
+  /// will give you a 1:1 mapping between logical pixels and the output pixels
+  /// in the image. The output image is not anti-aliased, so you may wish to
+  /// use a larger device pixel ratio to obtain an image large enough to
+  /// down sample.
+  ///
+  /// Defaults to 1.0.
+  double get pixelRatio => _pixelRatio;
+  double _pixelRatio = 1.0;
+  set pixelRatio(double pixelRatio) {
+    _pixelRatio = pixelRatio;
+    handleMetricsChanged();
+  }
+
+  /// Determines the dimensions of the virtual screen that the diagram will
+  /// be drawn on, in logical units.
+  Size get screenDimensions => _screenDimensions;
+  Size _screenDimensions = _kDefaultDiagramViewportSize;
+  set screenDimensions(Size screenDimensions) {
+    _screenDimensions = screenDimensions;
+    handleMetricsChanged();
+  }
+
+  /// Returns the singleton for the binding, after making sure it is
+  /// initialized.
+  static WidgetsBinding instance() {
+    if (WidgetsBinding.instance == null) {
+      new _DiagramFlutterBinding();
+    }
+    return WidgetsBinding.instance;
+  }
+
+  @override
+  ViewConfiguration createViewConfiguration() {
+    return new _DiagramViewConfiguration(
+      pixelRatio: pixelRatio,
+      size: screenDimensions,
+    );
+  }
+
+  /// Captures an image of the [RepaintBoundary] with the given key.
+  Future<ui.Image> takeSnapshot() async {
+    RenderRepaintBoundary object = _boundaryKey.currentContext.findRenderObject();
+    return await object.toImage(pixelRatio: pixelRatio);
+  }
+
+  /// Updates the current diagram with the given builder as the child of the
+  /// root widget, and generates a new frame.
+  void updateDiagram(
+    WidgetBuilder builder, {
+    Duration duration: Duration.zero,
+  }) {
+    Widget rootWidget = new _Diagram(
+      boundaryKey: _boundaryKey,
+      child: new Builder(builder: builder),
+    );
+    attachRootWidget(rootWidget);
+    pump();
+  }
+
+  /// Advances time by the given duration, and generates a new frame.
+  ///
+  /// The [duration] must not be null, or less than [Duration.zero].
+  void pump({Duration duration: Duration.zero}) {
+    assert(duration != null);
+    assert(duration >= Duration.zero);
+    _timestamp += duration;
+
+    handleBeginFrame(_timestamp);
+    handleDrawFrame();
```

Comment:
if you're driving the framework directly, do you even need to override the onBeginFrame? I'm confused as to your intent here.